### PR TITLE
Actually use the `#[do_not_recommend]` attribute if present

### PR DESCRIPTION
**Pull request by weiznich**
This change tweaks the error message generation to actually use the `#[do_not_recommend]` attribute if present by just skipping the marked trait impl in favour of the parent impl. It also adds a compile test for this behaviour. Without this change the test would output the following error:

```
error[E0277]: the trait bound `&str: Expression` is not satisfied
  --> /home/weiznich/Documents/rust/rust/tests/ui/diagnostic_namespace/do_not_recommend.rs:53:15
   |
LL |     SelectInt.check("bar");
   |               ^^^^^ the trait `Expression` is not implemented for `&str`, which is required by `&str: AsExpression<Integer>`
   |
   = help: the following other types implement trait `Expression`:
             Bound<T>
             SelectInt
note: required for `&str` to implement `AsExpression<Integer>`
  --> /home/weiznich/Documents/rust/rust/tests/ui/diagnostic_namespace/do_not_recommend.rs:26:13
   |
LL | impl<T, ST> AsExpression<ST> for T
   |             ^^^^^^^^^^^^^^^^     ^
LL | where
LL |     T: Expression<SqlType = ST>,
   |        ------------------------ unsatisfied trait bound introduced here
```

Note how that mentions `&str: Expression` before and now mentions `&str: AsExpression<Integer>` instead which is much more helpful for users.

Open points for further changes before stabilization:

* We likely want to move the attribute to the `#[diagnostic]` namespace to relax the guarantees given?
* How does it interact with the new trait solver?

r? @estebank

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (weiznich) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:474]: I believe that I need to change something with the obligation here to set the updated predicate so that the error label stops mentioning `Expression` as well. I just couldn't figure out how to do that yet. Help would be welcome here.

**Comment by weiznich** [comment]
The tests seem to pass locally with `./x.py test`. Any idea why might be the problem that causes the to fail on CI?

**Comment by compiler-errors** [comment]
Can you open a separate PR to move `do_not_recommend` to the diagnostic namespace? There's no reason for it to be a top-level attribute now that we have it.

**Comment by compiler-errors** [comment]
> How does it interact with the new trait solver?

I'll put up a PR to implement `do_not_recommend` in the new solver. It's basically a one-liner in the new solver.

**Comment by compiler-errors** [comment]
Also, this line should fix your ICEs:

https://github.com/rust-lang/rust/pull/124717/files#diff-09c366d3ad3ec9a42125253b610ca83cad6b156aa2a723f6c7e83eddef7b1e8fR521

I believe you should enable debug assertions in your local compiler:

(`config.toml`)
```
rust.debug = true
```

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:468]: This doesn't take into account obligation stacks that are like `A -requires> B -requires> C` where `A` and `B` are `#[do_not_recommend]` but `C` is not. I believe we need to basically need to pop `#[do_not_recommend]`s until there are none left in the stack *anywhere*, even if it isn't the *last* obligation in the stack.

For example, consider we put `#[do_not_recommend]` on this `IntoIterator` blanket, and then consider the following code:

```
trait OtherTrait {}

struct It<T>(T);
impl<T> Iterator for It<T> where T: OtherTrait {
    type Item = ();
    fn next(&mut self) -> Option<Self::Item> { todo!() }
}

fn main() {
    for x in It(()) {}
}
```

I would not expect the compiler to mention `(): OtherTrait`, but it would in this case, right? IMO it should just mention `It<()>: IntoIterator`.

From another perspective, considering the full tree of solving a goal like `It<()>: IntoIterator`, we simply wouldn't walk into the obligations that come from considering the blanket implementation marked with `#[do_not_recommend]`.

Comment (weiznich) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:468]: Doesn't that loop not already do that?

We reset the `oblication.cause.code()` field in line 475, so we should pop these obligations as long as we encounter ones that are marked with `#[do_not_recommend]`, right?

**Comment by weiznich** [comment]
> Can you open a separate PR to move do_not_recommend to the diagnostic namespace? There's no reason for it to be a top-level attribute now that we have it.

I will open a PR for that next week. That PR should keep the `#![feature(do_not_recommend)]` but just move the attribute to `#[diagnostic::on_unimplemented]`, right?

> Also, this line should fix your ICEs:
> 
> https://github.com/rust-lang/rust/pull/124717/files#diff-09c366d3ad3ec9a42125253b610ca83cad6b156aa2a723f6c7e83eddef7b1e8fR521

I rebased the PR to include your change. Hopefully that fixes the issue.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:468]: Not if the top obligation is not a ImplDerivedObligation, right?

**Comment by compiler-errors** [comment]
You messed up your commit history and have a bunch of merges. Please rebase them out by doing `git pull --rebase origin master` where origin is whatever git remote corresponds to rust-lang/rust (not your local fork)

**Comment by compiler-errors** [comment]
Actually, sorry, now that I look on this on my own web browser (not phone) I see there was a force-push to the master branch. Please do still rebase to fix that, though.

**Comment by weiznich** [comment]
> Actually, sorry, now that I look on this on my own web browser (not phone) I see there was a force-push to the master branch. Please do still rebase to fix that, though.

Whatever I managed to do there…
I've cleaned up that mess now.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (weiznich) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:468]: Thanks for the clarification. That sounds like I should first do some reading on what's actually represented by the `Obligation` struct. Is there any resource on this other than the source code?

Comment (weiznich) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs]: Hopefully it's now closer to what you imagined, although I'm still not sure about the details of `Obligation`.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (weiznich) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs]: That's resolved now

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:469]: Can you run `./x.py fmt`? This might not be the only thing that CI will complain about.

```suggestion
                            loop {
```

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:469]: It's probably that rustfmt bailed in this method because of something else, like a string that's too long or a where clause somewhere that it can't handle, because otherwise tidy would have failed.

Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:469]: Probably the `matches!`

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:485]: Can we add a comment in front of this? When skimming this code it will be really hard to figure out what's going on without paying a lot of attention.

**Review by estebank — APPROVED** [review]

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:470]: I don't think it's correct to use `peel_derives_with_predicate` here. We need to be peeling these one at a time.

Consider a where clause where the first impl we have to go through isn't marked with `#[do_not_recommend]` but the second is.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:475]: Let's not skip the binder here -- it will almost certainly cause an ICE later on. [`PolyTraitPredicate<'tcx>: ToPredicate<'tcx>`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/trait.ToPredicate.html#impl-ToPredicate%3C'tcx%3E-for-Binder%3C'tcx,+TraitPredicate%3C'tcx%3E%3E) should do the trick.

```suggestion
                                        obligation.predicate = base.to_predicate(tcx);
```

**Review by compiler-errors — CHANGES_REQUESTED** [review]
I'd like to take another look at this later before it lands.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:485]: It really should be extracted into a helper function, I think.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (weiznich) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs]: Hopefully it's more correct with https://github.com/rust-lang/rust/compare/eca373a073222883772da9c30616b3c746e3d6f2..f0279abd052e1ec4fa39ab71edbd7484dbc84c7b now. If that's not the case I would appreciate bit more guidance of how to address that.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (weiznich) [reply compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs]: I've fixed the formatting and extracted it into a helper function in https://github.com/rust-lang/rust/compare/eca373a073222883772da9c30616b3c746e3d6f2..f0279abd052e1ec4fa39ab71edbd7484dbc84c7b

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:1020]: nit:

```suggestion
                let code = (*c.derived.parent_code).clone();
```

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:1006]: This `'b` lifetime is unnecessary. It should always be `'tcx`. If not, then there's almost certainly some other lifetime in the caller that needs to be turned into a `'tcx`.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:1028]: Why does this need to be a separate loop? The logic that applies above (we need to peel until we find the obligation preceding the *first* `do_not_recommend` in the stack) should also apply below -- should be able to just have one loop.

**Review by compiler-errors — COMMENTED** [review]
I'd like to see more tests that exercise the cases that I've alluded to previously in review, since I'm not totally certain if the loop you've implemented is right :thinking:. For example, here's one test that I'd like to see works correctly:

```
#![feature(do_not_recommend)]

trait Root {}
trait DontRecommend {}
trait Other {}

#[do_not_recommend]
impl<T> Root for T where T: DontRecommend {}

impl<T> DontRecommend for T where T: Other {}

fn needs_root<T: Root>() {}

fn main() {
    needs_root::<()>();
    // Should not mention `(): Other` or `(): DontRecommend`
}
```

I'd also like to see if you can think of any other tests...

You should also be able to compare the tests against the implementation in the new solver by using revisions, tagging your tests with:

```rust
//@ revisions: current next
//@ ignore-compare-mode-next-solver (explicit revisions)
//@[next] compile-flags: -Znext-solver
```

They should look the same.

**Comment by weiznich** [comment]
Thanks for the review and the suggestions. I've hopefully applied everything of that, although I'm not sure about the errors from the new solver for one of the test cases.

File: tests/ui/diagnostic_namespace/do_not_recommend/as_expression.next.stderr
Comment (weiznich) [inline tests/ui/diagnostic_namespace/do_not_recommend/as_expression.next.stderr:1]: It seems like the new solver does not resolve associated types here yet?

File: tests/ui/diagnostic_namespace/do_not_recommend/as_expression.next.stderr
Comment (weiznich) [inline tests/ui/diagnostic_namespace/do_not_recommend/as_expression.next.stderr:19]: This seems to be an unrelated issue in the new solver?

File: tests/ui/diagnostic_namespace/do_not_recommend/stacked.current.stderr
Comment (compiler-errors) [inline tests/ui/diagnostic_namespace/do_not_recommend/stacked.current.stderr:7]: Yeah, see, this is what I mentioned. We are recommending a do_not_recommend trait 😅

File: tests/ui/diagnostic_namespace/do_not_recommend/as_expression.next.stderr
Comment (compiler-errors) [reply tests/ui/diagnostic_namespace/do_not_recommend/as_expression.next.stderr:19]: Yeah unrelated

File: tests/ui/diagnostic_namespace/do_not_recommend/stacked.current.stderr
Comment (weiznich) [reply tests/ui/diagnostic_namespace/do_not_recommend/stacked.current.stderr:7]: Hmm, yes.
It's not clear for me how to actually walk the tree here, as the `ObligationCause` seems to be the `ExprBindingObligation` variant in this case (according to a `dbg!()` call in the `apply_do_not_recommend` function). That version does not seem to contain any parent tree information, as far as I understand.

Comment (weiznich) [reply tests/ui/diagnostic_namespace/do_not_recommend/stacked.current.stderr]: Turns out that the problem was not the loop, but the location where I called the loop. It should work now as expected.

**Comment by compiler-errors** [comment]
Cool! I promise to validate this this weekend. Thanks for dealing with the review back-and-forth and fixing this.

**Review by estebank — APPROVED** [review]

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
@weiznich: `to_predicate` has been renamed to `uplift` in a recent commit of mine. Should be a quick fix, so I will just push it here and re-r+ it.

**Comment by compiler-errors** [comment]
Ah, tests need blessing too. I'll leave this up to you then. r=me when rebased.

@bors r-

**Comment by weiznich** [comment]
I've adjusted the tests. It should work now.

I will try to open a PR for moving the attribute to the `#[diagnostic]` namespace next week after this is merged. The branch for that already exists, but I need to rebase it on top of the final version accepted here.

**Pull request merged**
